gym/envs/registration.py

```diff
@@ -169,48 +174,166 @@ def __init__(self):
+
+            # Check if the package is installed
+            # If not instruct the user to install the package and then how to instantiate the env
+            if importlib.util.find_spec(relocated_package) is None:
+                message += f" Please install the package via `pip install {relocated_package}`."
+
+            # Otherwise the user should be able to instantiate the environment directly
+            if namespace != relocated_namespace:
+                message += f" You can instantiate the new namespaced environment as `{relocated_namespace}/{name}`."
+        # If the environment hasn't been relocated we'll construct a generic error message
+        else:
+            message = f"Environment `{name}` doesn't exist"
+            if namespace is not None:
+                message += f" in namespace `{namespace}`"
+            message += "."
+            suggestions = difflib.get_close_matches(name, self.names(namespace), n=1)
+            if suggestions:
+                message += f" Did you mean: `{suggestions[0]}`?"
+        # Throw the error
+        raise error.NameNotFound(message)
+
+    def _assert_version_exists(
+        self, namespace: Optional[str], name: str, version: Optional[int]
+    ):
+        self._assert_name_exists(namespace, name)
+        if version in self.tree[namespace][name]:
+            return
+
+        # Construct the appropriate exception.
+        # If the version is less than the latest version
+        # then we throw an error.DeprecatedEnv exception.
+        # Otherwise we throw error.VersionNotFound.
+        all_versions = self.tree[namespace][name].values()
+        latest_spec = max(all_versions, key=lambda spec: spec.version, default=None)
+
+        if version is not None:
+            message = f"Environment version `v{version}` for `"
+        else:
+            message = "The default version for `"
+
+        if namespace is not None:
+            message += f"{namespace}/"
+        message += f"{name}` "
+
+        # If this version doesn't exist but there exists a later version
+        # we should warn the user this version is deprecated.
         if (
-            namespace in self.tree
-            and name in self.tree[namespace]
-            and version in self.tree[namespace][name]
+            latest_spec
+            and latest_spec.version is not None
+            and version is not None
+            and version < latest_spec.version
         ):
-            return True
+            message += "is deprecated. "
+            message += f"Please use the latest version `v{latest_spec.version}`."
+            raise error.DeprecatedEnv(message)
+        # Otherwise we've asked for a version that doesn't exist.
         else:
-            return False
+            message += f"{name} could not be found. "
+            if all_versions:
+                all_versions_sorted = sorted(
```

Comment:
Also dangerous and is solved with the above changes to `_assert_version_exists`.